modules/secure-preview/token.php

```diff
@@ -0,0 +1,122 @@
+			return new WP_Error( 'vip-workflow-token-permission-denied', __( 'You do not have sufficient permissions to access this page.', 'vip-workflow' ) );
+		}
+
+		$token_lifetime = self::get_token_lifetime_in_seconds( $expiration_seconds );
+		$expiration     = time() + $token_lifetime;
+		$secret         = wp_generate_password( 64, true, true );
+		$token          = hash_hmac( 'sha256', self::TOKEN_ACTION, $secret );
+
+		$meta_value = [
+			'action'       => self::TOKEN_ACTION,
+			'expiration'   => $expiration,
+			'one_time_use' => $is_one_time_use,
+			'user'         => get_current_user_id(), // not validated, just stored in case it's interesting.
+			'token'        => $token,
+			'version'      => 1,                     // not validated, but might be useful in future.
+		];
+
+		add_post_meta( $post_id, self::META_KEY, $meta_value );
+
+		return $token;
+	}
+
+	/**
+	 * Validate a token that has been sent with a request.
+	 *
+	 * @param  string $token   Token.
+	 * @param  int    $post_id Post ID.
+	 * @return bool
+	 */
+	public static function validate_token( $token, $post_id ) {
+		$meta_key     = self::META_KEY;
+		$saved_tokens = get_post_meta( $post_id, $meta_key );
+		$current_time = time();
+		$is_valid     = false;
+
+		foreach ( $saved_tokens as $saved ) {
+			$is_expired = false;
+
+			// Check for token expiration.
+			if ( ! isset( $saved['expiration'] ) || $current_time > $saved['expiration'] ) {
+				$is_expired = true;
+			}
+
+			// Check if token matches. If it does, mark as expired.
+			if ( ! $is_expired && $token === $saved['token'] && self::TOKEN_ACTION === $saved['action'] ) {
+				$expire_on_use = ! isset( $saved['one_time_use'] ) || true === $saved['one_time_use'];
+
+				if ( $expire_on_use ) {
+					$is_expired = true;
+				}
+
+				$is_valid = true;
+			}
+
+			// Delete expired tokens.
+			if ( $is_expired ) {
+				delete_post_meta( $post_id, $meta_key, $saved );
+			}
+		}
+
+		return $is_valid;
+	}
+
+	/**
+	 * Get token lifetime (expiration period) in seconds.
+	 *
+	 * @param int  $token_lifetime_seconds The number of seconds this token should be valid for.
+	 * @return int
+	 */
+	private static function get_token_lifetime_in_seconds( $token_lifetime_seconds ) {
+		$default_max_lifetime_seconds = DAY_IN_SECONDS;
+
+		/**
+		 * Filter the maximum allowed token lifetime.
+		 *
+		 * This filter allows external plugins or custom code to modify the maximum
+		 * lifetime of a token. It provides a way to enforce an upper limit on the
+		 * duration for which a token remains valid.
+		 *
+		 * Understand the security implications of this change:
+		 * the longer the token lifetime, the more security risk you bear with a compromised token.
+		 * Extend the maximum lifetime with caution.
+		 *
+		 * @param int    $default_max_lifetime_seconds The default maximum token lifetime in seconds.
+		 */
+		$max_lifetime_seconds = apply_filters( 'vw_secure_preview_max_token_lifetime', $default_max_lifetime_seconds );
```

Comment:
I'm not sure I really see the point of a maximum, filterable or not, if all of the possible options are filterable via `vw_secure_preview_expiration_options`. It feels like this entire method is superfluous

Reply:
I agree! Removed in https://github.com/Automattic/vip-workflow-plugin/pull/19/commits/a624afa583a113fdedfd0804ac0c64f1edee9b1c.